apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/HashValidatingDownloader.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright 2016 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Download;
+using Google.Apis.Services;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+
+namespace Google.Cloud.Storage.V1
+{
+    /// <summary>
+    /// Subclass of <see cref="MediaDownloader"/> which validates the data it receives
+    /// against a CRC32c hash set in the header.
+    /// </summary>
+    internal sealed class HashValidatingDownloader : TmpMediaDownloader
+    {
+        private string crc32cHashBase64;
+        private Crc32c hasher;
+
+        /// <summary>Constructs a new downloader with the given client service.</summary>
+        internal HashValidatingDownloader(IClientService service) : base(service)
+        {
+        }
+
+        protected override void OnResponseReceived(HttpResponseMessage response)
+        {
+            base.OnResponseReceived(response);
+
+            crc32cHashBase64 = null;
+            hasher = new Crc32c();
```

Comment:
It might make sense to only create this when the header value is found and then null check in OnDataReceived. Although if the downloader can be reused, it might not be a good idea because then it would need to be reset back to null in various cases.